runtimes/core/src/metrics/registry.rs

```diff
@@ -0,0 +1,182 @@
+use crate::metrics::counter::{CounterOps, CounterSchemaBuilder};
+use crate::metrics::gauge::{GaugeOps, GaugeSchemaBuilder};
+
+use super::system::SystemMetricsCollector;
+use super::{Counter, Gauge};
+use dashmap::DashMap;
+use malachite::base::num::basic::traits::One;
+use metrics::{Key, Label};
+use std::sync::atomic::AtomicU64;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+struct MetricStorage {
+    atomic: Arc<AtomicU64>,
+    getter: Box<dyn Fn() -> MetricValue + Send + Sync>,
+    registered_at: SystemTime,
+}
+
+impl std::fmt::Debug for MetricStorage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MetricStorage")
+            .field("atomic", &self.atomic)
+            .field("registered_at", &self.registered_at)
+            .finish()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum MetricValue {
+    // Counter variants
+    CounterU64(u64),
+    CounterI64(i64),
+
+    // Gauge variants
+    GaugeU64(u64),
+    GaugeI64(i64),
+    GaugeF64(f64),
+}
+
+#[derive(Debug, Clone)]
+pub struct CollectedMetric {
+    pub key: Key,
+    pub value: MetricValue,
+    pub registered_at: SystemTime,
+}
+
+#[derive(Debug, Clone)]
```

Comment:
It's a bit unexpected to me that the registry itself is clonable. Would it be better to instead use `Arc<Registry>` in callers?

Reply:
yeah, that is nicer, updated to that now